api/populars.js

```diff
@@ -0,0 +1,93 @@
+const express = require("express");
+var cacheManager = require("cache-manager");
+var memoryCache = cacheManager.caching({
+  store: "memory",
+  max: 1000,
+  ttl: 86400 /*1 day in seconds*/
+});
+
+var router = express.Router();
+
+const { format, startOfDay, subMonths, subWeeks } = require("date-fns");
+
+function getDate({ period = "month", previous = 1 } = {}) {
+  if (period === "week") {
+    return format(subWeeks(startOfDay(new Date()), previous), "yyyy-MM-dd");
+  }
+  return format(subMonths(startOfDay(new Date()), previous), "yyyy-MM-dd");
+}
+
+function buildUrl(params) {
+  const qs = Object.entries(params).map(([key, value]) => `${key}=${value}`);
+  return `https://matomo.fabrique.social.gouv.fr/index.php?${qs.join("&")}`;
+}
+
+function mergePages(data) {
+  return data
+    .find(rootPage => rootPage.label === "fiche-service-public")
+    .subtable.concat(
+      data.find(rootPage => rootPage.label === "fiche-ministere-travail")
+        .subtable,
+      data.find(rootPage => rootPage.label === "contribution").subtable
+    );
+}
+
+const params = {
+  module: "API",
+  format: "Json",
+  idSite: 4,
+  method: "Actions.getPageUrls",
+  expanded: 1
+};
+
+async function getPopulars(period) {
+  const [data, previousPeriodData] = await Promise.all([
+    fetch(
+      buildUrl({ ...params, period, date: getDate({ period }) })
+    ).then(data => data.json()),
+    fetch(
+      buildUrl({
+        ...params,
+        period,
+        date: getDate({ period, previous: 2 })
+      })
+    ).then(data => data.json())
+  ]);
+
+  const candidates = mergePages(data).filter(page => page.label !== "Others");
+  const previousCandidates = mergePages(previousPeriodData);
+
+  const poundedCandidates = candidates.map(
+    ({ label, sum_daily_nb_uniq_visitors: views, url }) => {
+      const {
+        sum_daily_nb_uniq_visitors: previousViews
+      } = previousCandidates.find(
+        previousCandidate => candidates.url === previousCandidate.url
+      );
+      const growth = views / previousViews;
+      const score = Math.pow(growth, 2) * views;
```

Comment:
what is that score formula 
score = Views<sup>3</sup> / previousViews<sup>2</sup>

Reply:
le score c'est le (ratio vue cette semaine / semaine précédente) ^ 2 * nombre de vues.
En gros ce qui compte le plus c'est la croissance (qu'elle soit > 1 ou pas) mais on prend quand même en compte le nombre de vues dans le score histoire que la croissance de page pas très consultées ne soit pas trop déterminante

Reply:
hum ton n'explication est différente du code ;) 
ton code: 
```
nb de vue cette semaine / nb de semaine précédente * nb de vue cette semaine / nb de semaine précédente  * nb de vue cette semaine 
```

ce que je comprend de ton explication: 
```
nb de vue cette semaine / nb de semaine précédente * nb de vue cette semaine / nb de semaine précédente * nb de vue total
```

Reply:
`views` c'est le nombre de vues de la semaine et `previousViews` c'est le nombre de vue de la semaine précédente.
Je fais bien la distinction entre le ratio (que je mets au carré donc) et le nombre de vues dans la ligne de code qui calcule le `score` mais je ne suis pas certain de comprendre la différences que tu pointes ?
Ce serait plus clair si je renome `views` en `currentViews` ? J'imagine qu'il y a un truc qui ne doit pas être clair dans le code mais comme c'est moi qui l'ai écrit j'ai du mal à comprendre le soucis

Reply:
non non le naming est ok. j'aurais plutôt pondérer le `views` dans le calcul du score pour éviter d'avoir des écarts trop important par rapport au ratio qui de mon point de vue à plus de poids que le nombre de vues. mais bon on pourra voir ca plus tard